sycl/include/sycl/handler.hpp

```diff
@@ -1028,6 +1054,21 @@ class __SYCL_EXPORT handler {
                                 KernelType KernelFunc) {
     throwIfActionIsCreated();
     throwOnLocalAccessorMisuse<KernelName, KernelType>();
+
+    // Aside from the OpenCL CPU backend, the other backends cannot
+    // handle kernel launches when one or more of the dimensions of
+    // the global range exceeds the 32-bit max (*). However, in sycl,
+    // the global range size is specified by a size_t, which can be up
+    // to 64 bits. So, when we get such a global range, we wrap the
+    // old kernel in a new kernel that has each work item peform
+    // multiple invocations the old kernel in a 32-bit global range.
+#ifdef __SYCL_ENABLE_LARGE_RANGE_WRAPPING__
```

Comment:
This wrapping makes us conformant with the SYCL 2020 specification. Therefore, I would suggest to maybe flip the macro and have it enabled by default with an option to opt-out.

Perhaps @Pennycook or @gmlueck can also advise here. I guess that the wrapper kernel will always be emitted by the device compiler, because the `if` below is not `constexpr`. Therefore, enabling this creates a one more copy of a simple `parallel_for` kernel similar to range rounding mechanism below, which in sum brings us to 3 copies of the same kernel with 2 of them being potentially unused.

I remember @Pennycook suggesting that we integrate this wrapping into range rounding mechanism, but the latter is not a functional thing and it is indeed optional. Some users are disabling it to reduce code size, because they don't have performance issues the mechanism tries to solve. Brining in a functional aspect may not be the best solution.

Maybe we should do the wrapping always? This is a functionally required thing anyway and it will allow us not to generate a third copy of each `parallel_for(range)` kernel.

Reply:
My preference here would actually be to go one step further, and collapse things down to just one copy of `parallel_for(range)`.  You could view it as integrating the range rounding into this new wrapping mechanism, if that's more palatable. :smile:

Both this new wrapper and the range-rounding have to do two things:
1) Launch an `nd_range` that has a different size to the number of work-items requested by the user.
2) Ensure that if the `nd_range` is bigger than the user's requested size, certain iterations are skipped.

So, if we make this wrapped kernel the default, it would be straightforward to have it cover the range-rounding case too.  The `nd_range` would need to be rounded up before it's passed to the new wrapper kernel, but that could happen in host code.